File: data/derived_cot/rq3_prompting/results/translation/qwen/java_to_py/code_output/DatabaseProcessor.py
import sqlite3
from typing import List, Dict, Optional

class DatabaseProcessor:
    def __init__(self, database_name: str):
        self.database_name = database_name

    def create_table(self, table_name: str, key1: str, key2: str) -> None:
        with sqlite3.connect(self.database_name) as conn:
            cursor = conn.cursor()
            query = f"CREATE TABLE IF NOT EXISTS {table_name} (id INTEGER PRIMARY KEY, {key1} TEXT, {key2} INTEGER)"
            cursor.execute(query)
            conn.commit()

    def insert_into_database(self, table_name: str, data: List[Dict[str, object]]) -> None:
        with sqlite3.connect(self.database_name) as conn:
            cursor = conn.cursor()
            for item in data:
                name = item["name"]
                age = int(item["age"])  # Cast to int as in Java
                query = f"INSERT INTO {table_name} (name, age) VALUES ('{name}', {age})"
                cursor.execute(query)
            conn.commit()

    def search_database(self, table_name: str, name: str) -> Optional[List[Dict[str, object]]]:
        with sqlite3.connect(self.database_name) as conn:
            cursor = conn.cursor()
            query = f"SELECT * FROM {table_name} WHERE name = '{name}'"
            cursor.execute(query)
            columns = [col[0] for col in cursor.description]
            results = cursor.fetchall()
            if not results:
                return None
            rows = []
            for row in results:
                row_dict = dict(zip(columns, row))
                rows.append(row_dict)
            return rows

    def delete_from_database(self, table_name: str, name: str) -> None:
        with sqlite3.connect(self.database_name) as conn:
            cursor = conn.cursor()
            query = f"DELETE FROM {table_name} WHERE name = '{name}'"
            cursor.execute(query)
            conn.commit()

File: data/derived_cot/rq3_prompting/results/translation/r1/java_to_cpp/code_output/DatabaseProcessor.cpp
#include <sqlite3.h>
#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <variant>
#include <optional>
#include <sstream>

class DatabaseProcessor {
private:
    std::string databaseName;

    // Helper to execute a query that produces no result set
    void executeNonQuery(const std::string& query) {
        sqlite3* db = nullptr;
        char* errMsg = nullptr;
        int rc = sqlite3_open(databaseName.c_str(), &db);
        if (rc != SQLITE_OK) {
            std::cerr << "SQLite error: " << sqlite3_errmsg(db) << std::endl;
            if (db) sqlite3_close(db);
            return;
        }
        rc = sqlite3_exec(db, query.c_str(), nullptr, nullptr, &errMsg);
        if (rc != SQLITE_OK) {
            std::cerr << "SQLite error: " << errMsg << std::endl;
            sqlite3_free(errMsg);
        }
        sqlite3_close(db);
    }

public:
    DatabaseProcessor(const std::string& databaseName)
        : databaseName(databaseName) {}

    void createTable(const std::string& tableName,
                     const std::string& key1,
                     const std::string& key2) {
        std::ostringstream query;
        query << "CREATE TABLE IF NOT EXISTS " << tableName
              << " (id INTEGER PRIMARY KEY, " << key1 << " TEXT, " << key2 << " INTEGER)";
        executeNonQuery(query.str());
    }

    void insertIntoDatabase(const std::string& tableName,
                            const std::vector<std::map<std::string, std::variant<int, std::string>>>& data) {
        sqlite3* db = nullptr;
        int rc = sqlite3_open(databaseName.c_str(), &db);
        if (rc != SQLITE_OK) {
            std::cerr << "SQLite error: " << sqlite3_errmsg(db) << std::endl;
            if (db) sqlite3_close(db);
            return;
        }

        for (const auto& item : data) {
            // Extract values assuming keys "name" and "age" exist
            std::string name = std::get<std::string>(item.at("name"));
            int age = std::get<int>(item.at("age"));

            std::ostringstream query;
            query << "INSERT INTO " << tableName << " (name, age) VALUES ('"
                  << name << "', " << age << ")";

            char* errMsg = nullptr;
            rc = sqlite3_exec(db, query.str().c_str(), nullptr, nullptr, &errMsg);
            if (rc != SQLITE_OK) {
                std::cerr << "SQLite error: " << errMsg << std::endl;
                sqlite3_free(errMsg);
            }
        }
        sqlite3_close(db);
    }

    std::optional<std::vector<std::map<std::string, std::variant<int, std::string>>>>
    searchDatabase(const std::string& tableName, const std::string& name) {
        sqlite3* db = nullptr;
        int rc = sqlite3_open(databaseName.c_str(), &db);
        if (rc != SQLITE_OK) {
            std::cerr << "SQLite error: " << sqlite3_errmsg(db) << std::endl;
            if (db) sqlite3_close(db);
            return std::nullopt;
        }

        std::ostringstream query;
        query << "SELECT * FROM " << tableName << " WHERE name = '" << name << "'";

        sqlite3_stmt* stmt = nullptr;
        rc = sqlite3_prepare_v2(db, query.str().c_str(), -1, &stmt, nullptr);
        if (rc != SQLITE_OK) {
            std::cerr << "SQLite error: " << sqlite3_errmsg(db) << std::endl;
            sqlite3_close(db);
            return std::nullopt;
        }

        std::vector<std::map<std::string, std::variant<int, std::string>>> result;
        while (sqlite3_step(stmt) == SQLITE_ROW) {
            std::map<std::string, std::variant<int, std::string>> row;
            row["id"] = sqlite3_column_int(stmt, 0);
            row["name"] = reinterpret_cast<const char*>(sqlite3_column_text(stmt, 1));
            row["age"] = sqlite3_column_int(stmt, 2);
            result.push_back(row);
        }
        sqlite3_finalize(stmt);
        sqlite3_close(db);

        if (result.empty()) {
            return std::nullopt;  // mimic Java returning null
        }
        return result;
    }

    void deleteFromDatabase(const std::string& tableName, const std::string& name) {
        std::ostringstream query;
        query << "DELETE FROM " << tableName << " WHERE name = '" << name << "'";
        executeNonQuery(query.str());
    }
};